### DISCUSSION TITLE
Random midi CC

### DISCUSSION THREAD

**Discussion by djPtrvch** [tttapa/Control-Surface#1082] [Q&A]
Hello! I want to make a midi controller with 64 potentiometers for iPad. I use Arduino pro micro and 4 multiplexers cd74hc4067. 
My code. 
```cpp
// Array of mux
CD74HC4067 muxes[] = {
{
  A0,
   {0, 1, 2, 3},
},
{
  A1,
   {0, 1, 2, 3},
},
 {
   A2,
   {0, 1, 2, 3},
},
 {
   A3,
   {0, 1, 2, 3},
}
};

// Array of pots
CCPotentiometer potentiometers[] = {
    { muxes[0].pin(0), { 0, CHANNEL_13 } },
    { muxes[0].pin(1), { 1, CHANNEL_13 } },
    { muxes[0].pin(2), { 2, CHANNEL_13 } },
    { muxes[0].pin(3), { 3, CHANNEL_13 } },
    { muxes[0].pin(4), { 4, CHANNEL_13 } },
    { muxes[0].pin(5), { 5, CHANNEL_13 } },
    { muxes[0].pin(6), { 6, CHANNEL_13 } },
    { muxes[0].pin(7), { 7, CHANNEL_13 } },
    { muxes[0].pin(8), { 8, CHANNEL_13 } },
    { muxes[0].pin(9), { 9, CHANNEL_13 } },
    { muxes[0].pin(10), { 10, CHANNEL_13 } },
    { muxes[0].pin(11), { 11, CHANNEL_13 } },
    { muxes[0].pin(12), { 12, CHANNEL_13 } },
    { muxes[0].pin(13), { 13, CHANNEL_13 } },
    { muxes[0].pin(14), { 14, CHANNEL_13 } },
    { muxes[0].pin(15), { 15, CHANNEL_13 } },
 

    { muxes[1].pin(0), { 16, CHANNEL_13 } },
    { muxes[1].pin(1), { 17, CHANNEL_13 } },
    { muxes[1].pin(2), { 18, CHANNEL_13 } },
    { muxes[1].pin(3), { 19, CHANNEL_13 } },
    { muxes[1].pin(4), { 20, CHANNEL_13 } },
    { muxes[1].pin(5), { 21, CHANNEL_13 } },
    { muxes[1].pin(6), { 22, CHANNEL_13 } },
    { muxes[1].pin(7), { 23, CHANNEL_13 } },
    { muxes[1].pin(8), { 24, CHANNEL_13 } },
    { muxes[1].pin(9), { 25, CHANNEL_13 } },
    { muxes[1].pin(10), { 26, CHANNEL_13 } },
    { muxes[1].pin(11), { 27, CHANNEL_13 } },
    { muxes[1].pin(12), { 28, CHANNEL_13 } },
    { muxes[1].pin(13), { 29, CHANNEL_13 } },
    { muxes[1].pin(14), { 30, CHANNEL_13 } },
    { muxes[1].pin(15), { 31, CHANNEL_13 } },

    { muxes[2].pin(0), { 32, CHANNEL_13 } },
    { muxes[2].pin(1), { 33, CHANNEL_13 } },
    { muxes[2].pin(2), { 34, CHANNEL_13 } },
    { muxes[2].pin(3), { 35, CHANNEL_13 } },
    { muxes[2].pin(4), { 36, CHANNEL_13 } },
    { muxes[2].pin(5), { 37, CHANNEL_13 } },
    { muxes[2].pin(6), { 38, CHANNEL_13 } },
    { muxes[2].pin(7), { 39, CHANNEL_13 } },
    { muxes[2].pin(8), { 40, CHANNEL_13 } },
    { muxes[2].pin(9), { 41, CHANNEL_13 } },
    { muxes[2].pin(10), { 42, CHANNEL_13 } },
    { muxes[2].pin(11), { 43, CHANNEL_13 } },
    { muxes[2].pin(12), { 44, CHANNEL_13 } },
    { muxes[2].pin(13), { 45, CHANNEL_13 } },
    { muxes[2].pin(14), { 46, CHANNEL_13 } },
    { muxes[2].pin(15), { 47, CHANNEL_13 } },

    { muxes[3].pin(0), { 48, CHANNEL_13 } },
    { muxes[3].pin(1), { 49, CHANNEL_13 } },
    { muxes[3].pin(2), { 50, CHANNEL_13 } },
    { muxes[3].pin(3), { 51, CHANNEL_13 } },
    { muxes[3].pin(4), { 52, CHANNEL_13 } },
    { muxes[3].pin(5), { 53, CHANNEL_13 } },
    { muxes[3].pin(6), { 54, CHANNEL_13 } },
    { muxes[3].pin(7), { 55, CHANNEL_13 } },
    { muxes[3].pin(8), { 56, CHANNEL_13 } },
    { muxes[3].pin(9), { 57, CHANNEL_13 } },
    { muxes[3].pin(10), { 58, CHANNEL_13 } },
    { muxes[3].pin(11), { 59, CHANNEL_13 } },
    { muxes[3].pin(12), { 60, CHANNEL_13 } },
    { muxes[3].pin(13), { 61, CHANNEL_13 } },
    { muxes[3].pin(14), { 62, CHANNEL_13 } },
    { muxes[3].pin(15), { 63, CHANNEL_13 } },
};
```

When connected to iPad, Arduino sends a whole series of random midi CC. Tell me if I have an error in the code or somewhere else? In the test version, I only connected one multiplexer and a potentiometer to Arduino.

**Comment by tttapa** [comment]
> In the test version, I only connected one multiplexer and a potentiometer to Arduino.

It sounds like most of the inputs are floating, so they'll be picking up random noise. Your code should only contain potentiometers and muxes that are actually connected.

**Reply by djPtrvch** [reply]
I tried connecting one mux and four potentiometers. // Array of mux
CD74HC4067 muxes[] = {
{
A0,
{0, 1, 2, 3},
}
};

// Array of pots
CCPotentiometer potentiometers[] = {
{ muxes[0].pin(0), { 0, CHANNEL_13 } },
{ muxes[0].pin(1), { 1, CHANNEL_13 } },
{ muxes[0].pin(2), { 2, CHANNEL_13 } },
{ muxes[0].pin(3), { 3, CHANNEL_13 } },

}; 

But when mapping to iPad CC messages continue to change randomly

**Reply by djPtrvch** [reply]
> > In the test version, I only connected one multiplexer and a potentiometer to Arduino.
> 
> It sounds like most of the inputs are floating, so they'll be picking up random noise. Your code should only contain potentiometers and muxes that are actually connected.

Please tell me where the error might be

**Comment by djPtrvch** [comment]
![Uploading PXL_20241020_075811094.jpg…]()

**Reply by tttapa** [reply]
It looks like you posted the reply before the image upload was complete.

**Reply by djPtrvch** [reply]
Apparently yes. Re-uploaded.

**Comment by djPtrvch** [comment]
![PXL_20241020_075811094](https://github.com/user-attachments/assets/44820bf1-8c0d-4cd5-9324-e8d7ccf7902a)

**Reply by tttapa** [reply]
Please post your full code. The MIDI messages shown here cannot have originated from the code you posted in https://github.com/tttapa/Control-Surface/discussions/1082#discussioncomment-10994829 (e.g. the channels are all wrong).

**Reply by djPtrvch** [reply]
/** 
 * This example demonstrates the use of MIDI Control Change potentiometers that
 * can be used for changing effect parameters, volumes, pan and balance
 * controls, etc. It can control almost any knob in your DAW software. 
 * This example shows how to define multiple potentiometers using an array.
 *
 * @boards  AVR, AVR USB, Nano Every, Due, Nano 33 IoT, Nano 33 BLE, UNO R4, Teensy 3.x
 * 
 * Connections
 * -----------
 * 
 * - A0: wiper of the first potentiometer
 * - A1: wiper of the second potentiometer
 * - ...
 * - A5: wiper of the sixth potentiometer
 * 
 * Connect the left terminal of the potentiometers to ground, and the right one
 * to V<sub>CC</sub>.
 * 
 * Behavior
 * --------
 * 
 * - When you turn a potentiometer, you should receive MIDI Control Change
 *   events, with a value between 0 and 127.
 * - The analog inputs are filtered, so there shouldn't be any noise on the 
 *   position. If there is, check your wiring, and make sure that the resistance
 *   of the potentiometers isn't too high (10 kΩ is ideal).
 * 
 * Mapping
 * -------
 * 
 * Select the Arduino as a custom MIDI controller in your DAW, and use the 
 * MIDI learn option to assign the potentiometers to a function.
 * 
 * Written by PieterP, 2019-08-13  
 * https://github.com/tttapa/Control-Surface
 */

#include <Control_Surface.h> // Include the Control Surface library

// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;

// Array of mux
CD74HC4067 muxes[] = {
{
  A0,
   {0, 1, 2, 3},
}
};

// Array of pots
CCPotentiometer potentiometers[] = {
    { muxes[0].pin(0), { 0, CHANNEL_13 } },
    { muxes[0].pin(1), { 1, CHANNEL_13 } },
    { muxes[0].pin(2), { 2, CHANNEL_13 } },
    { muxes[0].pin(3), { 3, CHANNEL_13 } },

};


void setup() {
  Control_Surface.begin(); // Initialize Control Surface
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
}

**Comment by djPtrvch** [comment]
Could this be a problem with the Arduino board itself?

**Comment by djPtrvch** [comment]
I tried connecting potentiometers to Arduino without a multiplexer, same thing

**Comment by djPtrvch** [comment]
@tttapa Please tell me where the error might be

**Reply by LukeThomasDawson** [reply]
A couple of things to try:
(1)  For every input that you define, make sure it actually has a valid input physically connected to the corresponding pin.  So if you define 4 CCPotentiometers, make sure you have 4 pots connected up to those pins so that none of the pins are left "floating"
(2) Try creating just one CCPotentiometer and connecting one physical pot up to that pin.  Then do the same thing but specify a different pin.  Then you can test your setup one pin/function at a time.

**Reply by djPtrvch** [reply]
> A couple of things to try: (1) For every input that you define, make sure it actually has a valid input physically connected to the corresponding pin. So if you define 4 CCPotentiometers, make sure you have 4 pots connected up to those pins so that none of the pins are left "floating" (2) Try creating just one CCPotentiometer and connecting one physical pot up to that pin. Then do the same thing but specify a different pin. Then you can test your setup one pin/function at a time.

I did as you advised and connected one potentiometer separately to each output. Everything works. But as soon as I connected 4 potentiometers to 4 analog outputs, the CC values began to jump.

**Reply by tttapa** [reply]
What's the resistance of your potentiometers? If it's too large, you get cross-talk between ADC channels.

**Reply by djPtrvch** [reply]
> What's the resistance of your potentiometers? If it's too large, you get cross-talk between ADC channels.

10kOm
